core/include/detray/propagator/navigator.hpp

```diff
@@ -438,231 +515,174 @@ class navigator {
+                navigation.volume() != navigation.current()->link
+                    ? navigation::status::e_on_portal
+                    : navigation::status::e_on_module,
+                navigation.current()->index, navigation::trust_level::e_full);
+        } else {
+            // Otherwise the track is moving towards a surface
             navigation.set_state(navigation::status::e_towards_object,
                                  dindex_invalid,
                                  navigation::trust_level::e_full);
-        } else {
-            navigation.set_state(navigation::status::e_unknown, dindex_invalid,
-                                 navigation::trust_level::e_no_trust);
         }
-        // Release stepping constraints from actors
-        // Call the inspector before returning (only for debugging)
-        navigation.run_inspector("Set next: ");
-    }
-
-    /// Helper method to check and perform a volume switch
-    ///
-    /// @param navigation is the navigation state
-    DETRAY_HOST_DEVICE
-    bool check_volume_switch(state &navigation) const {
-        // Check if we need to switch volume index and (re-)initialize
-        // Do this when we are on a portal that has a volume link different to
-        // the volume we currently navigate in
-        if (navigation.status() == navigation::status::e_on_target and
-            navigation.volume() != navigation.current()->link) {
-            // Set volume index to the next volume provided by the portal
-            navigation.set_volume(navigation.current()->link);
-            // Initialize the new volume in the next update call
+        // Generally happens when after an update no next candidate in the
+        // cache is reachable anymore -> triggers init of [new] volume
+        if (navigation.is_exhausted()) {
             navigation.set_no_trust();
-
-            // We reached the end of the detector world
-            if (navigation.volume() == dindex_invalid) {
-                // heartbeat
-                return navigation.exit();
-            }
         }
-        return true;
     }
 
-    /// Helper method that updates the intersection of a single candidate
+    /// Helper method that updates the intersection of a single candidate and
+    /// checks reachability
     ///
-    /// @tparam track_t type of the track (including context)
+    /// @tparam track_t type of the track parametrization
     ///
+    /// @param candidate the intersection to be updated
     /// @param track the track information
+    ///
+    /// @returns whether the track can reach this candidate.
     template <typename track_t>
-    DETRAY_HOST_DEVICE inline void update_candidate(
-        const track_t &track, intersection_type &candidate) const {
+    DETRAY_HOST_DEVICE inline bool update_candidate(
+        intersection_type &candidate, const track_t &track) const {
+        // Remember the surface this candidate belongs to
         const dindex obj_idx = candidate.index;
```

Comment:
Not sure if this is still necessary

Reply:
Sadly it is :(. I want to really set the surface index in the ```intersection_kernel``` at some point.